### DISCUSSION TITLE
Implementing SpongeCake BSDF: Sampling half-vectors correctly.

### DISCUSSION THREAD

**Discussion by Vrroom** [mitsuba-renderer/mitsuba3#1167] [Q&A] [answered]
Hello Everyone,

I'm implementing the [SpongeCake BSDF](https://wangningbei.github.io/2022/SpongeCake_files/paper_SpongeCake.pdf). This is a BSDF that has a symmetric reflecting and transmitting Normal Distribution Function (NDF). I want to ensure that I'm sampling the half vector correctly. 

This is what I have so far:

```py
        alpha = dr.maximum(0.0001, self.alpha)
        S = dr.diag(mi.Vector3f(alpha * alpha, alpha * alpha, 1)) # surface type matrix. Later we'll rotate it and all that.
        sh_frame = mi.Frame3f(si.wi)
        # https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#sggx_sample, it seems that
        # the sh_frame is constructed as mi.Frame3f(si.wi). I checked via PDB that this is different from si.sh_frame
        # even if this is correct, the question remains whether the h is in the correct space
        h = sggx_sample(sh_frame, sample2, S)
        wo = mi.reflect(si.wi, h) 
```

My questions are whether I'm using: 

* The correct sh_frame to sample h. Should it be `mi.Frame3f(si.wi)` or `si.sh_frame`. I checked via PDB and the two are different. 
* What coordinate frame is the sampled h in. The docs referenced say that this is in the **world space** but based on my understanding of this stuff, I need an h in the **local shading frame**.

Happy to provide more information. Hope my question is clear. Thanks everyone. 

Sumit.

**Comment by ziyi-zhang** [comment]
Hi,
For surface interactions `si`, the shading frame is usually `mi.Frame3f(shading_normal)`. But for medium interactions `mei`, the shading frame is typically `mi.Frame3f(mei.wi)`.
The spongecake BSDF is a special case that entails both concepts: it assumes microflake volume under a surface. You need to be careful not to misuse the frame and the coordinate system. Also note that `si.wi` is the incident direction in surface normal's local frame, not in world space.
Best

**Comment by Vrroom** [comment, marked as answer]
Thanks Ziyi. I found the answer to my question but forgot to post it. This is slightly different from your comment. Let me write it here for anyone stumbling into this in the future. 

In the case of the SpongeCake BSDF, the phase function `f(wi, wo)` to be sampled is the specular phase function from the [SGGX microflake](https://research.nvidia.com/sites/default/files/pubs/2015-08_The-SGGX-microflake//sggx.pdf) (See Eq 24 in the linked in the paper). The algorithm sampling from this phase function requires constructing an orthonormal basis around `wi` (See Algorithm 1 in the SGGX paper).  This is why, the correct frame to pass to `sggx_sample` is `sh_frame = mi.Frame3f(si.wi)` as in my code above. 

In this case, I have erroneously named the orthonormal basis as `sh_frame`. It is not a shading frame at all. Rather it helps us compute and sample from the Visible (from the direction `wi`) NDF of the SGGX microflake.

**Reply by Vrroom** [reply]
Ok. Actually you did kind of say that when you mentioned `mei.wi`. My bad.
